src/wallet/wallet.cpp

```diff
@@ -6126,54 +6126,134 @@ bool PaymentAddressBelongsToWallet::operator()(const libzcash::SaplingPaymentAdd
 }
 bool PaymentAddressBelongsToWallet::operator()(const libzcash::UnifiedAddress &uaddr) const
 {
-    // TODO
+    auto hdChain = m_wallet->GetMnemonicHDChain();
+    assert(hdChain.has_value());
+
+    auto ufvkid = m_wallet->FindUnifiedFullViewingKey(uaddr);
+    if (ufvkid.has_value()) {
+        // Look through the UFVKs that we have generated, and confirm that the
+        // seed fingerprint for the key we find for the ufvkid corresponds to
+        // the wallet's mnemonic seed.
+        for (const auto& [k, v] : m_wallet->mapUnifiedAccountKeys) {
+            if (v == ufvkid.value() && k.first == hdChain.value().GetSeedFingerprint()) {
+                return true;
+            }
+        }
+    }
+
     return false;
 }
 
 // GetSourceForPaymentAddress
 
-PaymentAddressSource GetSourceForPaymentAddress::operator()(const CKeyID &zaddr) const
+PaymentAddressSource GetSourceForPaymentAddress::GetUnifiedSource(const libzcash::Receiver& receiver) const
 {
-    // TODO
-    return AddressNotFound;
+    auto hdChain = m_wallet->GetMnemonicHDChain();
+    assert(hdChain.has_value());
+
+    auto ufvkMeta = m_wallet->GetUFVKMetadataForReceiver(receiver);
+    if (ufvkMeta.has_value()) {
+        auto ufvkid = ufvkMeta.value().first;
+        // Look through the UFVKs that we have generated, and confirm that the
+        // seed fingerprint for the key we find for the ufvkid corresponds to
+        // the wallet's mnemonic seed.
+        for (const auto& [k, v] : m_wallet->mapUnifiedAccountKeys) {
+            if (v == ufvkid && k.first == hdChain.value().GetSeedFingerprint()) {
+                return PaymentAddressSource::MnemonicHDSeed;
+            }
+        }
+        return PaymentAddressSource::ImportedWatchOnly;
+    } else {
+        return PaymentAddressSource::AddressNotFound;
+    }
 }
-PaymentAddressSource GetSourceForPaymentAddress::operator()(const CScriptID &zaddr) const
+
+PaymentAddressSource GetSourceForPaymentAddress::operator()(const CKeyID &addr) const
 {
-    // TODO
-    return AddressNotFound;
+    auto ufvkSource = this->GetUnifiedSource(addr);
+    if (ufvkSource == PaymentAddressSource::AddressNotFound) {
+        if (m_wallet->HaveKey(addr)) {
+            return PaymentAddressSource::Random;
+        } else {
+            if (m_wallet->HaveWatchOnly(GetScriptForDestination(addr))) {
+                return PaymentAddressSource::ImportedWatchOnly;
+            }
+        }
+    }
+
+    return ufvkSource;
+}
+PaymentAddressSource GetSourceForPaymentAddress::operator()(const CScriptID &addr) const
+{
+    auto ufvkSource = this->GetUnifiedSource(addr);
+    if (ufvkSource == PaymentAddressSource::AddressNotFound) {
+        if (m_wallet->HaveCScript(addr)) {
+            return PaymentAddressSource::Imported;
```

Comment:
`listaddresses` does not currently use this payment source for transparent addresses. Both `CKeyID` and `CScriptId` are treated as `legacy_random` (using the heuristic mentioned above). If it is indeed the case that all scripts are imported and never generated from random data, then `listaddresses` should be altered to use this visitor as the source of truth.